plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/TransferSemaphoresHolder.java

```diff
@@ -0,0 +1,143 @@
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.repositories.s3.async;
+
+import org.opensearch.common.blobstore.stream.write.WritePriority;
+
+import java.util.Objects;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Transfer semaphore holder for controlled transfer of data to remote.
+ */
+public class TransferSemaphoresHolder {
+
+    // For tests
+    protected TypeSemaphore lowPrioritySemaphore;
+    protected TypeSemaphore highPrioritySemaphore;
+    private final int highPriorityPermits;
+    private final int lowPriorityPermits;
+    private final int acquireWaitDuration;
+    private final TimeUnit acquireWaitDurationUnit;
+
+    /**
+     * Constructor to create semaphores holder.
+     */
+    public TransferSemaphoresHolder(int availablePermits, double priorityPermitAllocation, int acquireWaitDuration, TimeUnit timeUnit) {
+
+        this.highPriorityPermits = (int) (priorityPermitAllocation * availablePermits);
+        this.highPrioritySemaphore = new TypeSemaphore(highPriorityPermits, "high");
+        this.lowPriorityPermits = availablePermits - highPriorityPermits;
+        this.lowPrioritySemaphore = new TypeSemaphore(lowPriorityPermits, "low");
+        this.acquireWaitDuration = acquireWaitDuration;
+        this.acquireWaitDurationUnit = timeUnit;
+    }
+
+    /**
+     * Overridden semaphore to identify transfer semaphores among all other semaphores for triaging.
+     */
+    public static class TypeSemaphore extends Semaphore {
+        private final String type;
+
+        public TypeSemaphore(int permits, String type) {
+            super(permits);
+            this.type = type;
+        }
+
+        public String getType() {
+            return type;
+        }
+    }
+
+    /**
+     * For multiple part requests of a single file, request context object will be set with the decision if low
+     * priority permits can also be utilized in high priority transfers of parts of the file. If high priority get fully
+     * consumed then low priority permits will be acquired for transfer.
+     *
+     * If a low priority transfer request comes in and a high priority transfer is in progress then till current
+     * high priority transfer finishes, low priority transfer may have to compete. This is an acceptable side effect
+     * because low priority transfers are generally heavy and it is ok to have slow progress in the beginning.
+     *
+     */
+    public static class RequestContext {
+
+        private final boolean lowPriorityPermitsConsumable;
+
+        private RequestContext(boolean lowPriorityPermitsConsumable) {
+            this.lowPriorityPermitsConsumable = lowPriorityPermitsConsumable;
+        }
+
+    }
+
+    public RequestContext createRequestContext() {
+        return new RequestContext(this.lowPrioritySemaphore.availablePermits() == lowPriorityPermits);
```

Comment:
Sorry i asked before as well , why do this need to be equal - does it mean normal priority can get low priority permits , when no low priority uploads are done ?